api/nodes_test.go

```diff
@@ -375,3 +376,160 @@ func TestNodes_GcAlloc(t *testing.T) {
 	require.NotNil(err)
 	require.True(structs.IsErrUnknownAllocation(err))
 }
+
+// Unittest monitorDrainMultiplex when an error occurs
+func TestNodes_MonitorDrain_Multiplex_Bad(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+
+	ctx := context.Background()
+	multiplexCtx, cancel := context.WithCancel(ctx)
+
+	// monitorDrainMultiplex doesn't require anything on *Nodes, so we
+	// don't need to use a full Client
+	var nodeClient *Nodes
+
+	outCh := make(chan string, 8)
+	errCh := make(chan string, 1)
+	nodeCh := make(chan string, 1)
+	allocCh := make(chan string, 8)
+	exitedCh := make(chan struct{})
+	go func() {
+		defer close(exitedCh)
+		nodeClient.monitorDrainMultiplex(ctx, cancel, outCh, errCh, nodeCh, allocCh)
+	}()
+
+	// Fake an alloc update
+	msg := "alloc update"
+	allocCh <- msg
+	require.Equal(msg, <-outCh)
+
+	// Fake a node update
+	msg = "node update"
+	nodeCh <- msg
+	require.Equal(msg, <-outCh)
+
+	// Fake an error that should shut everything down
+	msg = "fake error"
+	errCh <- msg
+	require.Equal(msg, <-outCh)
+
+	_, ok := <-exitedCh
+	require.False(ok)
+
+	_, ok = <-outCh
+	require.False(ok)
+
+	// Exiting should also cancel the context that would be passed to the
+	// node & alloc watchers
+	select {
+	case <-multiplexCtx.Done():
+	case <-time.After(100 * time.Millisecond):
+		t.Fatalf("context wasn't canceled")
+	}
+
+}
+
+// Unittest monitorDrainMultiplex when drain finishes
```

Comment:
Do you need `Unittest` specified here?

Reply:
It is pretty ugly huh? :\

I wanted to make it clear this wasn't actually testing the API calls to a real service+client and all that... basically just unittest the subtle chan multiplexing and error conditions.